stdlib/dynarray.ml

```diff
@@ -102,17 +53,246 @@ and 'a slot =
    The following conditions define what we call a "valid" dynarray:
    - valid length: [length <= Array.length arr]
    - no missing element in the live space:
-     forall i, [0 <= i < length] implies [arr.(i) <> Empty]
+     forall i, [0 <= i < length] implies [arr.(i) != dummy]
    - no element in the empty space:
-     forall i, [length <= i < Array.length arr] implies [arr.(i) = Empty]
+     forall i, [length <= i < Array.length arr] implies [arr.(i) == dummy]
 
    Unfortunately, we cannot easily enforce validity as an invariant in
    presence of concurrent updates. We can thus observe dynarrays in
    "invalid states". Our implementation may raise exceptions or return
    incorrect results on observing invalid states, but of course it
    must preserve memory safety.
+
+   {3 Dummies and flat float arrays}
+
+   OCaml performs a dynamic optimization of the representation of
+   float arrays, which is incompatible with our use of a dummy
+   value: if we initialize an array with user-provided elements,
+   it may get an optimized into a "flat float array", and
+   writing our non-float dummy into it would crash.
```

Comment:
It is really really hard to write a naked float into an array, and I think that this could in fact be considered a bug in the runtime that we managed to do it using `Array.blit`. I don't think it is worth mentioning here -- but I will go add an assertion in `caml_array_blit` against this.